// secretjs/dist/protobuf_stuff/secret/compute/v1beta1/query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebImpl = exports.QueryAddressByLabelDesc = exports.QueryLabelByAddressDesc = exports.QueryCodeHashByCodeIDDesc = exports.QueryCodeHashByContractAddressDesc = exports.QueryCodesDesc = exports.QueryCodeDesc = exports.QueryQuerySecretContractDesc = exports.QueryContractsByCodeIDDesc = exports.QueryContractInfoDesc = exports.QueryDesc = exports.QueryClientImpl = exports.DecryptedAnswer = exports.QueryCodeHashResponse = exports.QueryContractLabelResponse = exports.QueryContractAddressResponse = exports.QueryCodesResponse = exports.QueryCodeResponse = exports.CodeInfoResponse = exports.QueryContractsByCodeIDResponse = exports.ContractInfoWithAddress = exports.QueryContractInfoResponse = exports.QuerySecretContractResponse = exports.QueryByCodeIDRequest = exports.QueryByContractAddressRequest = exports.QueryByLabelRequest = exports.QuerySecretContractRequest = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const grpc_web_1 = require("@improbable-eng/grpc-web");
const _m0 = __importStar(require("protobufjs/minimal"));
const types_1 = require("../../../secret/compute/v1beta1/types");
const empty_1 = require("../../../google/protobuf/empty");
const browser_headers_1 = require("browser-headers");
const abci_1 = require("../../../cosmos/base/abci/v1beta1/abci");
exports.protobufPackage = "secret.compute.v1beta1";
function createBaseQuerySecretContractRequest() {
    return { contractAddress: "", query: new Uint8Array() };
}
exports.QuerySecretContractRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.contractAddress !== "") {
            writer.uint32(10).string(message.contractAddress);
        }
        if (message.query.length !== 0) {
            writer.uint32(18).bytes(message.query);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQuerySecretContractRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contractAddress = reader.string();
                    break;
                case 2:
                    message.query = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            contractAddress: isSet(object.contractAddress)
                ? String(object.contractAddress)
                : "",
            query: isSet(object.query)
                ? bytesFromBase64(object.query)
                : new Uint8Array(),
        };
    },
    toJSON(message) {
        const obj = {};
        message.contractAddress !== undefined &&
            (obj.contractAddress = message.contractAddress);
        message.query !== undefined &&
            (obj.query = base64FromBytes(message.query !== undefined ? message.query : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseQuerySecretContractRequest();
        message.contractAddress = (_a = object.contractAddress) !== null && _a !== void 0 ? _a : "";
        message.query = (_b = object.query) !== null && _b !== void 0 ? _b : new Uint8Array();
        return message;
    },
};
function createBaseQueryByLabelRequest() {
    return { label: "" };
}
exports.QueryByLabelRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.label !== "") {
            writer.uint32(10).string(message.label);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryByLabelRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.label = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            label: isSet(object.label) ? String(object.label) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.label !== undefined && (obj.label = message.label);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryByLabelRequest();
        message.label = (_a = object.label) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseQueryByContractAddressRequest() {
    return { contractAddress: "" };
}
exports.QueryByContractAddressRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.contractAddress !== "") {
            writer.uint32(10).string(message.contractAddress);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryByContractAddressRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contractAddress = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            contractAddress: isSet(object.contractAddress)
                ? String(object.contractAddress)
                : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.contractAddress !== undefined &&
            (obj.contractAddress = message.contractAddress);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryByContractAddressRequest();
        message.contractAddress = (_a = object.contractAddress) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseQueryByCodeIDRequest() {
    return { codeId: "0" };
}
exports.QueryByCodeIDRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.codeId !== "0") {
            writer.uint32(8).uint64(message.codeId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryByCodeIDRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.codeId = longToString(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            codeId: isSet(object.codeId) ? String(object.codeId) : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        message.codeId !== undefined && (obj.codeId = message.codeId);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryByCodeIDRequest();
        message.codeId = (_a = object.codeId) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
function createBaseQuerySecretContractResponse() {
    return { data: new Uint8Array() };
}
exports.QuerySecretContractResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.data.length !== 0) {
            writer.uint32(10).bytes(message.data);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQuerySecretContractResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.data = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            data: isSet(object.data)
                ? bytesFromBase64(object.data)
                : new Uint8Array(),
        };
    },
    toJSON(message) {
        const obj = {};
        message.data !== undefined &&
            (obj.data = base64FromBytes(message.data !== undefined ? message.data : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQuerySecretContractResponse();
        message.data = (_a = object.data) !== null && _a !== void 0 ? _a : new Uint8Array();
        return message;
    },
};
function createBaseQueryContractInfoResponse() {
    return { contractAddress: "", ContractInfo: undefined };
}
exports.QueryContractInfoResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.contractAddress !== "") {
            writer.uint32(10).string(message.contractAddress);
        }
        if (message.ContractInfo !== undefined) {
            types_1.ContractInfo.encode(message.ContractInfo, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryContractInfoResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contractAddress = reader.string();
                    break;
                case 2:
                    message.ContractInfo = types_1.ContractInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            contractAddress: isSet(object.contractAddress)
                ? String(object.contractAddress)
                : "",
            ContractInfo: isSet(object.ContractInfo)
                ? types_1.ContractInfo.fromJSON(object.ContractInfo)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.contractAddress !== undefined &&
            (obj.contractAddress = message.contractAddress);
        message.ContractInfo !== undefined &&
            (obj.ContractInfo = message.ContractInfo
                ? types_1.ContractInfo.toJSON(message.ContractInfo)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryContractInfoResponse();
        message.contractAddress = (_a = object.contractAddress) !== null && _a !== void 0 ? _a : "";
        message.ContractInfo =
            object.ContractInfo !== undefined && object.ContractInfo !== null
                ? types_1.ContractInfo.fromPartial(object.ContractInfo)
                : undefined;
        return message;
    },
};
function createBaseContractInfoWithAddress() {
    return { contractAddress: "", ContractInfo: undefined };
}
exports.ContractInfoWithAddress = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.contractAddress !== "") {
            writer.uint32(10).string(message.contractAddress);
        }
        if (message.ContractInfo !== undefined) {
            types_1.ContractInfo.encode(message.ContractInfo, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseContractInfoWithAddress();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contractAddress = reader.string();
                    break;
                case 2:
                    message.ContractInfo = types_1.ContractInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            contractAddress: isSet(object.contractAddress)
                ? String(object.contractAddress)
                : "",
            ContractInfo: isSet(object.ContractInfo)
                ? types_1.ContractInfo.fromJSON(object.ContractInfo)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.contractAddress !== undefined &&
            (obj.contractAddress = message.contractAddress);
        message.ContractInfo !== undefined &&
            (obj.ContractInfo = message.ContractInfo
                ? types_1.ContractInfo.toJSON(message.ContractInfo)
                : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseContractInfoWithAddress();
        message.contractAddress = (_a = object.contractAddress) !== null && _a !== void 0 ? _a : "";
        message.ContractInfo =
            object.ContractInfo !== undefined && object.ContractInfo !== null
                ? types_1.ContractInfo.fromPartial(object.ContractInfo)
                : undefined;
        return message;
    },
};
function createBaseQueryContractsByCodeIDResponse() {
    return { contractInfos: [] };
}
exports.QueryContractsByCodeIDResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.contractInfos) {
            exports.ContractInfoWithAddress.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryContractsByCodeIDResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contractInfos.push(exports.ContractInfoWithAddress.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            contractInfos: Array.isArray(object === null || object === void 0 ? void 0 : object.contractInfos)
                ? object.contractInfos.map((e) => exports.ContractInfoWithAddress.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.contractInfos) {
            obj.contractInfos = message.contractInfos.map((e) => e ? exports.ContractInfoWithAddress.toJSON(e) : undefined);
        }
        else {
            obj.contractInfos = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryContractsByCodeIDResponse();
        message.contractInfos =
            ((_a = object.contractInfos) === null || _a === void 0 ? void 0 : _a.map((e) => exports.ContractInfoWithAddress.fromPartial(e))) || [];
        return message;
    },
};
function createBaseCodeInfoResponse() {
    return { codeId: "0", creator: "", codeHash: "", source: "", builder: "" };
}
exports.CodeInfoResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.codeId !== "0") {
            writer.uint32(8).uint64(message.codeId);
        }
        if (message.creator !== "") {
            writer.uint32(18).string(message.creator);
        }
        if (message.codeHash !== "") {
            writer.uint32(26).string(message.codeHash);
        }
        if (message.source !== "") {
            writer.uint32(34).string(message.source);
        }
        if (message.builder !== "") {
            writer.uint32(42).string(message.builder);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCodeInfoResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.codeId = longToString(reader.uint64());
                    break;
                case 2:
                    message.creator = reader.string();
                    break;
                case 3:
                    message.codeHash = reader.string();
                    break;
                case 4:
                    message.source = reader.string();
                    break;
                case 5:
                    message.builder = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            codeId: isSet(object.codeId) ? String(object.codeId) : "0",
            creator: isSet(object.creator) ? String(object.creator) : "",
            codeHash: isSet(object.codeHash) ? String(object.codeHash) : "",
            source: isSet(object.source) ? String(object.source) : "",
            builder: isSet(object.builder) ? String(object.builder) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.codeId !== undefined && (obj.codeId = message.codeId);
        message.creator !== undefined && (obj.creator = message.creator);
        message.codeHash !== undefined && (obj.codeHash = message.codeHash);
        message.source !== undefined && (obj.source = message.source);
        message.builder !== undefined && (obj.builder = message.builder);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseCodeInfoResponse();
        message.codeId = (_a = object.codeId) !== null && _a !== void 0 ? _a : "0";
        message.creator = (_b = object.creator) !== null && _b !== void 0 ? _b : "";
        message.codeHash = (_c = object.codeHash) !== null && _c !== void 0 ? _c : "";
        message.source = (_d = object.source) !== null && _d !== void 0 ? _d : "";
        message.builder = (_e = object.builder) !== null && _e !== void 0 ? _e : "";
        return message;
    },
};
function createBaseQueryCodeResponse() {
    return { codeInfo: undefined, wasm: new Uint8Array() };
}
exports.QueryCodeResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.codeInfo !== undefined) {
            exports.CodeInfoResponse.encode(message.codeInfo, writer.uint32(10).fork()).ldelim();
        }
        if (message.wasm.length !== 0) {
            writer.uint32(18).bytes(message.wasm);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryCodeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.codeInfo = exports.CodeInfoResponse.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.wasm = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            codeInfo: isSet(object.codeInfo)
                ? exports.CodeInfoResponse.fromJSON(object.codeInfo)
                : undefined,
            wasm: isSet(object.wasm)
                ? bytesFromBase64(object.wasm)
                : new Uint8Array(),
        };
    },
    toJSON(message) {
        const obj = {};
        message.codeInfo !== undefined &&
            (obj.codeInfo = message.codeInfo
                ? exports.CodeInfoResponse.toJSON(message.codeInfo)
                : undefined);
        message.wasm !== undefined &&
            (obj.wasm = base64FromBytes(message.wasm !== undefined ? message.wasm : new Uint8Array()));
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryCodeResponse();
        message.codeInfo =
            object.codeInfo !== undefined && object.codeInfo !== null
                ? exports.CodeInfoResponse.fromPartial(object.codeInfo)
                : undefined;
        message.wasm = (_a = object.wasm) !== null && _a !== void 0 ? _a : new Uint8Array();
        return message;
    },
};
function createBaseQueryCodesResponse() {
    return { codeInfos: [] };
}
exports.QueryCodesResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.codeInfos) {
            exports.CodeInfoResponse.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryCodesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.codeInfos.push(exports.CodeInfoResponse.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            codeInfos: Array.isArray(object === null || object === void 0 ? void 0 : object.codeInfos)
                ? object.codeInfos.map((e) => exports.CodeInfoResponse.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.codeInfos) {
            obj.codeInfos = message.codeInfos.map((e) => e ? exports.CodeInfoResponse.toJSON(e) : undefined);
        }
        else {
            obj.codeInfos = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryCodesResponse();
        message.codeInfos =
            ((_a = object.codeInfos) === null || _a === void 0 ? void 0 : _a.map((e) => exports.CodeInfoResponse.fromPartial(e))) || [];
        return message;
    },
};
function createBaseQueryContractAddressResponse() {
    return { contractAddress: "" };
}
exports.QueryContractAddressResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.contractAddress !== "") {
            writer.uint32(10).string(message.contractAddress);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryContractAddressResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.contractAddress = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            contractAddress: isSet(object.contractAddress)
                ? String(object.contractAddress)
                : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.contractAddress !== undefined &&
            (obj.contractAddress = message.contractAddress);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryContractAddressResponse();
        message.contractAddress = (_a = object.contractAddress) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseQueryContractLabelResponse() {
    return { label: "" };
}
exports.QueryContractLabelResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.label !== "") {
            writer.uint32(10).string(message.label);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryContractLabelResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.label = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            label: isSet(object.label) ? String(object.label) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.label !== undefined && (obj.label = message.label);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryContractLabelResponse();
        message.label = (_a = object.label) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseQueryCodeHashResponse() {
    return { codeHash: "" };
}
exports.QueryCodeHashResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.codeHash !== "") {
            writer.uint32(10).string(message.codeHash);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryCodeHashResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.codeHash = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            codeHash: isSet(object.codeHash) ? String(object.codeHash) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.codeHash !== undefined && (obj.codeHash = message.codeHash);
        return obj;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryCodeHashResponse();
        message.codeHash = (_a = object.codeHash) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseDecryptedAnswer() {
    return {
        type: "",
        input: "",
        outputData: "",
        outputDataAsString: "",
        outputLogs: [],
        outputError: new Uint8Array(),
        plaintextError: "",
    };
}
exports.DecryptedAnswer = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.type !== "") {
            writer.uint32(10).string(message.type);
        }
        if (message.input !== "") {
            writer.uint32(18).string(message.input);
        }
        if (message.outputData !== "") {
            writer.uint32(26).string(message.outputData);
        }
        if (message.outputDataAsString !== "") {
            writer.uint32(34).string(message.outputDataAsString);
        }
        for (const v of message.outputLogs) {
            abci_1.StringEvent.encode(v, writer.uint32(42).fork()).ldelim();
        }
        if (message.outputError.length !== 0) {
            writer.uint32(50).bytes(message.outputError);
        }
        if (message.plaintextError !== "") {
            writer.uint32(58).string(message.plaintextError);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDecryptedAnswer();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.type = reader.string();
                    break;
                case 2:
                    message.input = reader.string();
                    break;
                case 3:
                    message.outputData = reader.string();
                    break;
                case 4:
                    message.outputDataAsString = reader.string();
                    break;
                case 5:
                    message.outputLogs.push(abci_1.StringEvent.decode(reader, reader.uint32()));
                    break;
                case 6:
                    message.outputError = reader.bytes();
                    break;
                case 7:
                    message.plaintextError = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            type: isSet(object.type) ? String(object.type) : "",
            input: isSet(object.input) ? String(object.input) : "",
            outputData: isSet(object.outputData) ? String(object.outputData) : "",
            outputDataAsString: isSet(object.outputDataAsString)
                ? String(object.outputDataAsString)
                : "",
            outputLogs: Array.isArray(object === null || object === void 0 ? void 0 : object.outputLogs)
                ? object.outputLogs.map((e) => abci_1.StringEvent.fromJSON(e))
                : [],
            outputError: isSet(object.outputError)
                ? bytesFromBase64(object.outputError)
                : new Uint8Array(),
            plaintextError: isSet(object.plaintextError)
                ? String(object.plaintextError)
                : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.type !== undefined && (obj.type = message.type);
        message.input !== undefined && (obj.input = message.input);
        message.outputData !== undefined && (obj.outputData = message.outputData);
        message.outputDataAsString !== undefined &&
            (obj.outputDataAsString = message.outputDataAsString);
        if (message.outputLogs) {
            obj.outputLogs = message.outputLogs.map((e) => e ? abci_1.StringEvent.toJSON(e) : undefined);
        }
        else {
            obj.outputLogs = [];
        }
        message.outputError !== undefined &&
            (obj.outputError = base64FromBytes(message.outputError !== undefined
                ? message.outputError
                : new Uint8Array()));
        message.plaintextError !== undefined &&
            (obj.plaintextError = message.plaintextError);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseDecryptedAnswer();
        message.type = (_a = object.type) !== null && _a !== void 0 ? _a : "";
        message.input = (_b = object.input) !== null && _b !== void 0 ? _b : "";
        message.outputData = (_c = object.outputData) !== null && _c !== void 0 ? _c : "";
        message.outputDataAsString = (_d = object.outputDataAsString) !== null && _d !== void 0 ? _d : "";
        message.outputLogs =
            ((_e = object.outputLogs) === null || _e === void 0 ? void 0 : _e.map((e) => abci_1.StringEvent.fromPartial(e))) || [];
        message.outputError = (_f = object.outputError) !== null && _f !== void 0 ? _f : new Uint8Array();
        message.plaintextError = (_g = object.plaintextError) !== null && _g !== void 0 ? _g : "";
        return message;
    },
};
class QueryClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.contractInfo = this.contractInfo.bind(this);
        this.contractsByCodeID = this.contractsByCodeID.bind(this);
        this.querySecretContract = this.querySecretContract.bind(this);
        this.code = this.code.bind(this);
        this.codes = this.codes.bind(this);
        this.codeHashByContractAddress = this.codeHashByContractAddress.bind(this);
        this.codeHashByCodeID = this.codeHashByCodeID.bind(this);
        this.labelByAddress = this.labelByAddress.bind(this);
        this.addressByLabel = this.addressByLabel.bind(this);
    }
    contractInfo(request, metadata) {
        return this.rpc.unary(exports.QueryContractInfoDesc, exports.QueryByContractAddressRequest.fromPartial(request), metadata);
    }
    contractsByCodeID(request, metadata) {
        return this.rpc.unary(exports.QueryContractsByCodeIDDesc, exports.QueryByCodeIDRequest.fromPartial(request), metadata);
    }
    querySecretContract(request, metadata) {
        return this.rpc.unary(exports.QueryQuerySecretContractDesc, exports.QuerySecretContractRequest.fromPartial(request), metadata);
    }
    code(request, metadata) {
        return this.rpc.unary(exports.QueryCodeDesc, exports.QueryByCodeIDRequest.fromPartial(request), metadata);
    }
    codes(request, metadata) {
        return this.rpc.unary(exports.QueryCodesDesc, empty_1.Empty.fromPartial(request), metadata);
    }
    codeHashByContractAddress(request, metadata) {
        return this.rpc.unary(exports.QueryCodeHashByContractAddressDesc, exports.QueryByContractAddressRequest.fromPartial(request), metadata);
    }
    codeHashByCodeID(request, metadata) {
        return this.rpc.unary(exports.QueryCodeHashByCodeIDDesc, exports.QueryByCodeIDRequest.fromPartial(request), metadata);
    }
    labelByAddress(request, metadata) {
        return this.rpc.unary(exports.QueryLabelByAddressDesc, exports.QueryByContractAddressRequest.fromPartial(request), metadata);
    }
    addressByLabel(request, metadata) {
        return this.rpc.unary(exports.QueryAddressByLabelDesc, exports.QueryByLabelRequest.fromPartial(request), metadata);
    }
}
exports.QueryClientImpl = QueryClientImpl;
exports.QueryDesc = {
    serviceName: "secret.compute.v1beta1.Query",
};
exports.QueryContractInfoDesc = {
    methodName: "ContractInfo",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QueryByContractAddressRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryContractInfoResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryContractsByCodeIDDesc = {
    methodName: "ContractsByCodeID",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QueryByCodeIDRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryContractsByCodeIDResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryQuerySecretContractDesc = {
    methodName: "QuerySecretContract",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QuerySecretContractRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QuerySecretContractResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryCodeDesc = {
    methodName: "Code",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QueryByCodeIDRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryCodeResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryCodesDesc = {
    methodName: "Codes",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return empty_1.Empty.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryCodesResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryCodeHashByContractAddressDesc = {
    methodName: "CodeHashByContractAddress",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QueryByContractAddressRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryCodeHashResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryCodeHashByCodeIDDesc = {
    methodName: "CodeHashByCodeID",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QueryByCodeIDRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryCodeHashResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryLabelByAddressDesc = {
    methodName: "LabelByAddress",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QueryByContractAddressRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryContractLabelResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
exports.QueryAddressByLabelDesc = {
    methodName: "AddressByLabel",
    service: exports.QueryDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary() {
            return exports.QueryByLabelRequest.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary(data) {
            return Object.assign(Object.assign({}, exports.QueryContractAddressResponse.decode(data)), { toObject() {
                    return this;
                } });
        },
    },
};
class GrpcWebImpl {
    constructor(host, options) {
        this.host = host;
        this.options = options;
    }
    unary(methodDesc, _request, metadata) {
        var _a;
        const request = Object.assign(Object.assign({}, _request), methodDesc.requestType);
        const maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(Object.assign(Object.assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata || this.options.metadata;
        return new Promise((resolve, reject) => {
            grpc_web_1.grpc.unary(methodDesc, {
                request,
                host: this.host,
                metadata: maybeCombinedMetadata,
                transport: this.options.transport,
                debug: this.options.debug,
                onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message);
                    }
                    else {
                        const err = new Error(response.statusMessage);
                        err.code = response.status;
                        err.metadata = response.trailers;
                        reject(err);
                    }
                },
            });
        });
    }
}
exports.GrpcWebImpl = GrpcWebImpl;
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
const atob = globalThis.atob ||
    ((b64) => globalThis.Buffer.from(b64, "base64").toString("binary"));
function bytesFromBase64(b64) {
    const bin = atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
        arr[i] = bin.charCodeAt(i);
    }
    return arr;
}
const btoa = globalThis.btoa ||
    ((bin) => globalThis.Buffer.from(bin, "binary").toString("base64"));
function base64FromBytes(arr) {
    const bin = [];
    for (const byte of arr) {
        bin.push(String.fromCharCode(byte));
    }
    return btoa(bin.join(""));
}
function longToString(long) {
    return long.toString();
}
if (_m0.util.Long !== long_1.default) {
    _m0.util.Long = long_1.default;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
